Source/Core/MaxHolesLambdaLifter.cs

```diff
@@ -0,0 +1,392 @@
+/// variables.
+///
+/// To encode templates we store a <see cref="_templates"/> dictionary that maps each node to its template,
+/// an object of type <see cref="LambdaLiftingTemplate"/>.
+///
+/// The first phase traverses a lambda's AST bottom-up, pupulating the <see cref="_templates"/> map.
+/// At the end of the first phase, we obtain a list of replacement expressions that correspond to the holes in
+/// the lambda.
+///
+/// The second phase replaces the templates with bound variables. It is handled by the
+/// <see cref="LambdaLiftingMaxHolesFiller.Fill"/> method.
+///
+/// This class requires a lambda's `old` expressions to occur only around free variables (this
+/// transformation is performed in <c>LambdaHelper.LiftLambdaMaxHoles</c>). 
+/// </summary>
+class MaxHolesLambdaLifter : StandardVisitor {
+    private readonly List<Variable> _nestedBoundVariables = new List<Variable>();
+
+    private readonly LambdaExpr _lambda;
+    private readonly Dictionary<Expr, FunctionCall> _liftedLambdas;
+    private readonly String _freshFnName;
+    private readonly List<Function> _lambdaFunctions;
+    private readonly List<Expr> _lambdaAxioms;
+    private int _freshVarCount;
+
+    private readonly Dictionary<Absy, LambdaLiftingTemplate> _templates = new Dictionary<Absy, LambdaLiftingTemplate>();
+
+    public MaxHolesLambdaLifter(
+        LambdaExpr lambda,
+        Dictionary<Expr, FunctionCall> liftedLambdas,
+        string freshFnName,
+        List<Function> lambdaFunctions,
+        List<Expr> lambdaAxioms,
+        int freshVarCount = 0
+    ) {
+        _lambda = lambda;
+        _liftedLambdas = liftedLambdas;
+        _freshFnName = freshFnName;
+        _lambdaFunctions = lambdaFunctions;
+        _lambdaAxioms = lambdaAxioms;
+        _freshVarCount = freshVarCount;
+    }
+
+    private string FreshVarNameGen(List<string> existing) {
+        const string prefix = "l#";
+        while (true) {
+            var newName = prefix + _freshVarCount;
+            _freshVarCount++;
+            if (!existing.Contains(newName)) return newName;
+        }
+    }
+
+    private bool IsBound(Variable node) {
+        return _lambda.Dummies.Contains(node) || _nestedBoundVariables.Contains(node);
+    }
+
+    public override Expr VisitNAryExpr(NAryExpr node) {
+        Contract.Requires(node != null);
+        if (_templates.ContainsKey(node)) return node;
+        base.VisitNAryExpr(node);
+        var nodeArgs = node.Args;
+        if (nodeArgs.Any(arg => _templates[arg].ContainsBoundVariables())) {
+            var replacements = new List<Expr>();
+            foreach (Expr arg in nodeArgs) {
+                replacements.AddRange(_templates[arg].GetReplacements());
+            }
+
+            _templates[node] = new TemplateWithBoundVariables(replacements);
+        } else {
+            // todo (MR) Could we avoid the casting? Could we encode in type system that all children of hole nodes are holes?
```

Comment:
No, C#'s type system does not allow you to do this, unless you made a complete copy of the AST classes (which I don't recommend). But this is a good invariant to be stated in the AST class for hole nodes.